liblwgeom/lwgeom.c

```diff
@@ -1606,55 +1621,29 @@ lwgeom_remove_repeated_points_in_place(LWGEOM *geom, double tolerance)
 		}
 		case MULTIPOINTTYPE:
 		{
-			double tolsq = tolerance*tolerance;
-			uint32_t i, j, n = 0;
-			LWMPOINT *mpt = (LWMPOINT *)(geom);
-			LWPOINT **out;
-			LWPOINT *out_stack[out_stack_size];
-			int use_heap = (mpt->ngeoms > out_stack_size);
-
-			/* No-op on empty */
-			if (mpt->ngeoms < 2)
-				return geometry_modified;
-
-			/* We cannot write directly back to the multipoint */
-			/* geoms array because we're reading out of it still */
-			/* so we use a side array */
-			if (use_heap)
-				out = lwalloc(sizeof(LWMPOINT *) * mpt->ngeoms);
-			else
-				out = out_stack;
+			double tolsq = tolerance * tolerance;
+			LWMPOINT *mpt = (LWMPOINT *)geom;
+			qsort(mpt->geoms, mpt->ngeoms, sizeof(LWPOINT *), point_cmp);
 
-			/* Inefficient O(n^2) implementation */
-			for (i = 0; i < mpt->ngeoms; i++)
+			uint32_t i = 0;
+			for (uint32_t j = 1; j < mpt->ngeoms; j++)
 			{
-				int seen = 0;
-				LWPOINT *p1 = mpt->geoms[i];
-				const POINT2D *pt1 = getPoint2d_cp(p1->point, 0);
-				for (j = 0; j < n; j++)
+				POINT2D ptj, pti;
+				getPoint2d_p(mpt->geoms[i]->point, 0, &pti);
+				getPoint2d_p(mpt->geoms[j]->point, 0, &ptj);
+				if (distance2d_sqr_pt_pt(&ptj, &pti) > tolsq)
```

Comment:
This is not enough - for instance, in the square in the center of this image there can possibly be three points within the distance.
![image](https://user-images.githubusercontent.com/810638/115175513-cba40600-a0d3-11eb-8226-2c9eaeeaed31.png)

Your method can still be a special case for tolerance=0, just plug back in old one :)

I see three big options how to implement this:

I. via double grid snap:
 - snap to grid of size = tolerance
 - check all the ones in the same bucket if they are the same
 - snap to grid of size = tolerance with a shift of 0.5 
 - check all the ones in the same bucket if they are the same
 
II. double-sorting with window tracking:
 - order by X
 - iterate through list, look for duplicates back in the list until X is within tolerance
 - order by Y
 - iterate through list, look for duplicates back in the list until Y is within tolerance
 
III. tree
 - build a tree
 - while building a tree insertions, skip inserting points that have a point in tree within radius.